### ISSUE TITLE
How is the matrix U computed?

### ISSUE THREAD

**Issue by cam1681** [yangorwell/SENG-Optimizer#1]
I have read the paper of the SENG, but I didn't get how is the U (In equation (9)) computed? It is clear that U is the derivatives of the loss_i at different point i (which forms the i-th column of the matrix U).  If we use loss.backward to compute the grad, the only thing we can get is the derivatives of the total loss (namely \sum_i loss_i) with respect to the parameters. If we use autograd function in PyTorch to compute the U and it will be inefficient. Thus, how did you achieve the U without much cost? Thanks very much for solving my puzzles and thanks again for the good work you have committed!

**Comment by yangorwell** [comment]
Thanks for your attention to our work! Sorry to reply to you late.

To get the gradient with respect to each sample, please see the codes https://github.com/yangorwell/SENG/blob/66b7148a956c44a4fb523548a5e77d89e314bef8/Pytorch/imagenet/seng_dist.py#L208.
In our implementation, we define a new class ''LayerEx'' in Line 21-Line 35 of resnet_seng.py and then we can obtain the necessary information to compute sample gradient.

You can also refer to the package https://github.com/f-dangel/backpack. 

If you have any other question, feel free to let me know!
Best,
Minghan.

**Comment by cam1681** [comment]
> Thanks for your attention to our work! Sorry to reply to you late.
> 
> To get the gradient with respect to each sample, please see the codes
> 
> https://github.com/yangorwell/SENG/blob/66b7148a956c44a4fb523548a5e77d89e314bef8/Pytorch/imagenet/seng_dist.py#L208
> 
> .
> In our implementation, we define a new class ''LayerEx'' in Line 21-Line 35 of resnet_seng.py and then we can obtain the necessary information to compute sample gradient.
> You can also refer to the package https://github.com/f-dangel/backpack.
> 
> If you have any other question, feel free to let me know! Best, Minghan.

Thanks for your reply! I got it.
